### PR TITLE
Fixes for issues reported by Klockwork

### PR DESCRIPTION
**Pull request by ranj063**

### CODE REVIEW

File: sound/soc/sof/ipc3-pcm.c [hunk 1]
```diff
@@ -213,9 +213,22 @@ static int sof_ipc3_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
...
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
-	struct sof_dai_private_data *private = dai->private;
+	struct sof_dai_private_data *private;
 	struct snd_soc_dpcm *dpcm;
 
+	if (!dai) {
+		dev_err(component->dev, "%s: No DAI found with name %s\n", __func__,
+			rtd->dai_link->name);
+		return -EINVAL;
+	}
```
Comment (plbossart) [inline sound/soc/sof/ipc3-pcm.c:223]: can this happen? or is this a possible false alarm?

File: sound/soc/sof/ipc3-pcm.c [hunk 2]
```diff
@@ -213,9 +213,22 @@ static int sof_ipc3_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
...
+
+	private = dai->private;
+	if (!private) {
+		dev_err(component->dev, "%s: No private data found for DAI %s\n", __func__,
+			rtd->dai_link->name);
+		return -EINVAL;
+	}
+
```
Comment (plbossart) [inline sound/soc/sof/ipc3-pcm.c:231]: same here, we have tons of code that deferences dai->private, is this really a problem here?

File: sound/soc/sof/ipc3.c [hunk 3]
```diff
@@ -385,7 +385,8 @@ static int sof_ipc3_set_get_data(struct snd_sof_dev *sdev, void *data, size_t da
 	size_t offset = 0;
 	u8 *src, *dst;
 	u32 num_msg;
-	int ret, i;
+	int ret = 0;
+	int i;
```
Comment (plbossart) [inline sound/soc/sof/ipc3.c:389]: this one is valid indeed.

File: sound/soc/sof/sof-audio.c [hunk 4]
```diff
@@ -343,6 +343,9 @@ sof_setup_or_free_widgets_in_order(struct snd_sof_dev *sdev, struct snd_soc_dapm
 	struct snd_soc_dapm_widget *widget;
 	int ret, i;
 
+	if (!list)
+		return 0;
+
```
Comment (plbossart) [inline sound/soc/sof/sof-audio.c]: it's either
a) a false alarm - this is already tested in a macro
b) a real problem, the same code is used in soc-dapm.c

````c
	wlist = dapm_kcontrol_get_wlist(update->kcontrol);

	for_each_dapm_widgets(wlist, wi, w) {

#define for_each_dapm_widgets(list, i, widget)				\
	for ((i) = 0;							\
	     (i) < list->num_widgets && (widget = list->widgets[i]);	\
	     (i)++)

````

File: sound/soc/sof/ipc3.c [hunk 5]
```diff
@@ -385,7 +385,8 @@ static int sof_ipc3_set_get_data(struct snd_sof_dev *sdev, void *data, size_t da
 	size_t offset = 0;
 	u8 *src, *dst;
 	u32 num_msg;
-	int ret, i;
+	int ret = 0;
```
Comment (ujfalusi) [inline sound/soc/sof/ipc3.c:388]: I think this a bogus report.
ret will be assigned as `num_msg` is minimum 1, so the for() loop is executed at least once and ret got the return value of send message.

If I'm not mistaken (looking at the code on a phone).

File: sound/soc/sof/ipc3-pcm.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/ipc3-pcm.c:223]: this one is not possible but there's no way for klockwork to know that. Dai's are assigned names during topology loading and provate data also set at the same time. So, its better to have this check here.

File: sound/soc/sof/sof-audio.c (hunk 4 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: this is definitely a false alarm. the csame check is there in the caller function too.

File: sound/soc/sof/ipc3.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/sof/ipc3.c:388]: you're right but that's hard for static analysis to figure out that 
num_msg = DIV_ROUND_UP(msg_bytes, payload_size);
is > 0.

Better to add the init IMHO

File: sound/soc/sof/sof-audio.c (hunk 4 above)
Comment (plbossart) [reply sound/soc/sof/sof-audio.c]: let's remove it then and mark the issue as N/A

File: sound/soc/sof/ipc3-pcm.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/sof/ipc3-pcm.c:223]: ok let's keep it then

File: sound/soc/sof/ipc3-pcm.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/sof/ipc3-pcm.c:231]: let's keep this as well.

File: sound/soc/sof/sof-audio.c (hunk 4 above)
Comment (ranj063) [reply sound/soc/sof/sof-audio.c]: dropped now

**Review by plbossart — APPROVED** [review]

**Review by bardliao — APPROVED** [review]

**Pull request merged**